docs/adrs/0549-composite-run-steps.md

```diff
@@ -0,0 +1,275 @@
+    - uses: actions/checkout@v2
+      needs: [job1, job2]
+    - uses: user/composite@v1
+    - name: workflow step 1
+      run: echo hello world 3
+    - name: workflow step 2
+      run: echo hello world 4
+```
+Example Output
+```
+[npm installation output]
+[pip requirements output]
+echo hello world 3
+echo hello world 4
+```
+We add a token called "composite" which allows our Runner code to process composite actions. By invoking "using: composite", our Runner code then processes the "steps" attribute, converts this template code to a list of steps, and finally runs each run step sequentially. If any step fails and there are no `if` conditions defined, the whole composite action job fails. 
+
+### Inputs
+Example `user/composite/action.yml`:
+```
+using: 'composite' 
+inputs:
+  your_name:
+    description: 'Your name'
+    default: 'Ethan'
+steps: 
+  - run: echo hello ${{ inputs.your_name }}
+```
+
+Example `workflow.yml`:
+```
+...
+steps: 
+  - id: foo
+    uses: user/composite@v1
+    with:
+      your_name: "Octocat"
+  - run: echo hello ${{ steps.foo.inputs.your_name }} 2
+```
+
+Example Output:
+```
+hello Octocat
+Error
+```
+
+Each input variable in the composite action is only viewable in its own scope (unlike environment variables). As seen in the last line in the example output, in the workflow file, it will not have access to the action's `inputs` attribute.
+
+### Outputs
+Example `user/composite/action.yml`:
+```
+using: 'composite' 
+inputs:
+  your_name:
+    description: 'Your name'
+    default: 'Ethan'
+outputs:
+  bar: ${{ steps.my-step.my-output}}
+steps: 
+  - id: my-step
+    run: |
+      echo ::set-output name=my-output::my-value
+      echo hello ${{ inputs.your_name }} 
+```
+
+Example `workflow.yml`:
+```
+...
+steps: 
+  - id: foo
+    uses: user/composite@v1
+    with:
+      your_name: "Octocat"
+  - run: echo oh ${{ steps.foo.outputs.bar }} 
+```
+
+Example Output:
+```
+hello Octocat
+oh hello Octocat
+```
+Each of the output variables from the composite action is viewable from the workflow file that uses the composite action. In other words, every child action output(s) is viewable only by its parent using dot notation (ex `steps.foo.outputs.bar`).
+
+Moreover, the output ids are only accessible within the scope where it was defined. Note that in the example above, in our `workflow.yml` file, it should not have access to output id (i.e. `my-output`). For example, in the `workflow.yml` file, you can't run `foo.steps.my-step.my-output`.
+
+Our philosophy is we don't want the workflow author to need to know how the internal workings of the action work.
+
+### Context
+Similar to the workflow file, the composite action has access to the [same context objects](https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#contexts) (ex: `github`, `env`, `strategy`). 
+
+### Environment
+Example `user/composite/action.yml`:
+```
+using: 'composite' 
+env:
```

Comment:
let's consider starting without `env` at this level in the action.yml (extra complexity)

also we should not allow the secrets context inside action.yml. Should pass secrets through.

Reply:
also should we consider not allowing `env` _context_ inside action.yml

e.g. can't do the following inside action.yml:

```yaml
steps:
  - run: echo ${{ env.foo }}
```

...even though explicily set `env` flow to the processes.